File: src/conversationlog/ConversationLog.java
package conversationlog;

import lib.io.FastReader;

import java.io.FileNotFoundException;
import java.util.*;
import static conversationlog.ConversationLog.skipWords;

// https://open.kattis.com/problems/conversationlog

public class ConversationLog {

    static Set<String> skipWords = new HashSet<>();

    public static void main(String[] args) throws FileNotFoundException {

        StringBuilder file = new StringBuilder("src/conversationlog/in2");
        FastReader io = new FastReader(file);
        int numCases = io.nextInt();

        var map = new HashMap<String, User>();
        for (int i = 0; i < numCases; i++) {
            String words = io.nextLine();
            String[] arr = words.split("\\s");

            var name = arr[0];
            var list = new ArrayList<>(Arrays.asList(arr).subList(1, arr.length));
            var user = new User(name, list);

            if (map.containsKey(name)) {
                map.get(name).words.addAll(list);
            } else {
                map.put(name, user);
            }
        }
        for (User user : map.values()) {
            user.uniqueWords = new HashSet<>(user.words);
        }

        var keyWords = new ArrayList<Match>();
        var keySet   = new HashSet<String>();
        for (User user : map.values()) {
            for (String uniqueWord : user.uniqueWords) {
                if (keySet.contains(uniqueWord) || skipWords.contains(uniqueWord)) continue;

                int numberOfMatches = user.allSetsContainWord(map, uniqueWord);
                if (numberOfMatches > 0) {
                    Match m = new Match(uniqueWord, numberOfMatches);
                    keyWords.add(m);
                    keySet.add(uniqueWord);
                }
            }
        }
        if (keyWords.size() == 0) {
            System.out.println("ALL CLEAR");
            return;
        }
        Collections.sort(keyWords);
        for (Match match : keyWords) {
            System.out.println(match.word);
        }
    }
}

class User {
    String name;
    List<String> words;
    Set<String> uniqueWords;

    public User(String name, List<String> words) {
        this.name = name;
        this.words = words;
    }

    int allSetsContainWord(Map<String, User> map, String uniqueWord) {
        Collection<User> values = map.values();
        for (User user : values) {
            if (!user.uniqueWords.contains(uniqueWord)) {
                skipWords.add(uniqueWord);
                return 0;
            }
        }

        int cnt = 0;
        for (User user : values) {
            for (String word : user.words) {
                if (word.equals(uniqueWord)) {
                    cnt++;
                }
            }
        }
        return cnt;
    }

    @Override
    public String toString() {
        return name;
    }
}

class Match implements Comparable<Match> {
    String word;
    int numberOfMatches;

    public Match(String word, int numberOfMatches) {
        this.word = word;
        this.numberOfMatches = numberOfMatches;
    }

    @Override
    public int compareTo(Match o) {
        int i = o.numberOfMatches - this.numberOfMatches;
        if (i != 0) {
            return i;
        }
        return this.word.compareTo(o.word);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Match match = (Match) obj;
        return this.word.equals(match.word);
    }

    @Override
    public int hashCode() {
        return Objects.hash(word);
    }
}

File: src/conversationlog/conversationlog.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"sort"
	"strconv"
	"strings"
)

// https://open.kattis.com/problems/conversationlog

type user struct {
	name  string
	words []string
	set   map[string]bool
}

type Word struct {
	word string
	cnt  int
}

var db = make(map[string]*user)
var keyWords = make(map[string]bool)
var skipWords = make(map[string]bool)

func main() {
	io, _ := ioutil.ReadAll(os.Stdin)
	io = io[:len(io)-1] 	// remove newline at end

	file := strings.Split(string(io), "\n")
	cases, _ := strconv.Atoi(file[0])

	var sb strings.Builder

	// populate a map of users -> words
	for i := 1; i <= cases; i++ {
		line := strings.Fields(file[i])

		if val, ok := db[line[0]]; ok {
			val.words = append(val.words, line[1:]...)
		} else {
			db[line[0]] = &user{
				name:  line[0],
				words: line[1:],
			}
		}
	}

	// create a set of words for each user
	for _, usr := range db {
		usr.set = make(map[string]bool)
		for _, word := range usr.words {
			usr.set[word] = true
		}
	}

	if cases == 1 {
		for _, usr := range db {
			for word, _ := range usr.set {
				sb.WriteString(word)
			}
		}
		fmt.Print(sb.String())
		return
	}

	for _, usr := range db {
		for word, _ := range usr.set {
			if canSkipWord(word) || isKeyWord(word) {
				continue
			}

			// if all users share common word, add it to a set 'keyWords'
			if allSetsContainWord(word) {
				keyWords[word] = true
			}
		}
	}

	if len(keyWords) == 0 {
		fmt.Println("ALL CLEAR")
		return
	}

	// count the # of times the keywords are used across all users
	var solution []Word

	i := 0
	for word, _ := range keyWords {
		key := &Word{word: word, cnt: 0}
		solution = append(solution, *key)

		for _, usr := range db {
			for _, wrd := range usr.words {
				if key.word == wrd {
					solution[i].cnt++
				}
			}
		}
		i++
	}

	// sort descending (sort by number first, then alphabetical on tie)
	sort.Slice(solution, func(i, j int) bool {
		if solution[i].cnt == solution[j].cnt {
			return solution[i].word < solution[j].word
		}
		return solution[i].cnt > solution[j].cnt
	})

	for _, word := range solution {
		sb.WriteString(word.word + "\n")
	}

	fmt.Print(sb.String())
}

func allSetsContainWord(word string) bool {
	for _, usr := range db {
		if !usr.set[word] {
			skipWords[word] = true
			return false
		}
	}
	return true
}

func canSkipWord(word string) bool {
	return skipWords[word]
}

func isKeyWord(word string) bool {
	return keyWords[word]
}
